nix/tests.nix

```diff
@@ -164,6 +164,78 @@ let
             postgrest --dump-schema \
             | ${yq}/bin/yq -y .
       '';
+
+  coverage =
+    name: postgresql:
+    checkedShellScript
+      {
+        inherit name;
+        docs = "Run spec and io tests while collecting hpc coverage data.";
+        inRootDir = true;
+      }
+      ''
+        env="$(cat ${postgrest.env})"
+        export PATH="$env/bin:$PATH"
+
+        # clean up previous coverage data
+        rm -rf coverage && mkdir coverage
+        find . -type f -iname "*.tix" -delete
+
+        # build once before running all the tests
+        ${cabal-install}/bin/cabal v2-build ${devCabalOptions} --enable-tests all
+
+        # collect all tests in one file (./postgrest.tix)
+        ${withTmpDb postgresql} ${cabal-install}/bin/cabal v2-exec ${devCabalOptions} \
+          ${ioTestPython}/bin/pytest -- -v test/io-tests
+        ${withTmpDb postgresql} ${cabal-install}/bin/cabal v2-test ${devCabalOptions}
+
+        # collect all the tix files
+        ${ghc}/bin/hpc sum  --union --exclude=Paths_postgrest --output=.hpc/tests.tix ./*.tix
+
+        # prepare the overlay
+        touch test/coverage.overlay
+        ${ghc}/bin/hpc overlay --output=.hpc/overlay.tix test/coverage.overlay
+        ${ghc}/bin/hpc sum --union --output=.hpc/tests-overlay.tix .hpc/tests.tix .hpc/overlay.tix
+
+        # check nothing in the overlay is actually tested
+        ${ghc}/bin/hpc map --function=inv --output=.hpc/inverted.tix .hpc/tests.tix
+        ${ghc}/bin/hpc combine --function=sub \
+          --output=.hpc/check.tix .hpc/overlay.tix .hpc/inverted.tix
+        # returns zero exit code if any count="<non-zero>" lines are found, i.e.
+        # something is covered by both the overlay and the tests
+        if ${ghc}/bin/hpc report --xml .hpc/check.tix | ${gnugrep}/bin/grep -qP 'count="[^0]'
+        then
+          ${ghc}/bin/hpc markup --highlight-covered --destdir=coverage/overlay .hpc/overlay.tix || true
+          ${ghc}/bin/hpc markup --highlight-covered --destdir=coverage/check .hpc/check.tix || true
+          echo "ERROR: Something is covered by both the tests and the overlay:"
+          echo "file://$(pwd)/coverage/check/hpc_index.html"
+          exit 1
+        else
+          # prepare codecov json report
+          ${hpc-codecov}/bin/hpc-codecov --mix=.hpc --out=coverage/codecov.json .hpc/tests-overlay.tix
+
+          # create html and stdout reports
+          # TODO: The markup command fails when run outside nix-shell (i.e. in CI!)
+          # Need to fix it properly in the future instead of adding the || true
+          ${ghc}/bin/hpc markup --destdir=coverage .hpc/tests-overlay.tix || true
```

Comment:
This `|| true` is really ugly. Once `postgrest-coverage` is started from outside nix-shell by installing it with nix-env, it starts failing at this line with the following error:

```
Writing: Main.hs.html
hpc: ./main/Main.hs: hGetContents: invalid argument (invalid byte sequence)
```

I have no clue where this comes from. When I add `--exclude Main` to the hpc command it works, but of course the html file for Main is not written. So there seems to be something special about the `Main.hs` file - in combination with running the command outside nix-shell. @monacoremo maybe you can have a look? My nix skills are at their limits...